src/g_game.c

```diff
@@ -686,6 +686,18 @@ static void AdjustWeaponSelection(int *newweapon)
     // killough 2/8/98, 3/22/98 -- end of weapon selection changes
 }
 
+int G_GetNextWeapon(int direction)
+{
+    int result = G_NextWeapon(direction);
+
+    if (!demo_compatibility && !full_weapon_cycle)
+    {
+        AdjustWeaponSelection(&result);
+    }
```

Comment:
I'm struggling with this function. Any ideas?

Reply:
First, I suggest using "carousel" naming everywhere to avoid confusion with similarly named "next" weapon variables and functions.

If possible, the weapon carousel should be updated independently of the player's current or pending weapon. When the weapon carousel feature is enabled (it should be a menu toggle), a different weapon switching routine is used, meaning `next_weapon` is always zero.

The weapon carousel functions use their own previous weapon cycle and next weapon cycle, based on `demo_compatibility` and `full_weapon_cycle`. Maybe these are lists or arrays. They're updated when the player picks up a weapon, dies, etc.

When the carousel first pops up, the position is initially based on the player's current weapon. However, while the carousel is being cycled, the position is based on the desired target weapon, e.g. `carousel_weapon`, and NOT the player's current or pending weapon.

Every time the next or previous weapon input is pressed, the `carousel_weapon` variable is updated. An additional `else if` condition is added around [here](https://github.com/fabiangreffrath/woof/blob/58dcc9da3c8b137bb40ee2065f3c7c4edd970951/src/g_game.c#L911) to set `newweapon` to `carousel_weapon` if it's not `wp_nochange`.

This is like pressing the 1-9 keys directly. The weapon selection logic must be done ahead of time, mirroring what will happen when the tic is processed in G_BuildTiccmd and later in P_PlayerThink. This ensures that the icons don't mismatch.

There are probably some pitfalls to this approach. This is a tricky feature to get right.

Reply:
Wow, that is a lot of detail! I came up with `player_t local_player` variable to get code from `g_game.c` to work: 701d4ab Hacky, but it seems good enough.

Feel free to commit to this PR.

Reply:
> Wow, that is a lot of detail!

Yes, sorry. It comes from a desire to avoid the way the carousel gets "stuck" when switching weapons quickly from one direction to the next (e.g. press "next" a few times, then immediately press "prev" a few times and the carousel is unresponsive). For comparison, turn on weapon slots for a moment and notice how it's always immediately responsive to weapon switching. That's how I'd like to see the next/prev inputs behave as well.

Reply:
So this may be out of scope for this PR (next/prev weapon improvements).

Reply:
> So this may be out of scope for this PR (next/prev weapon improvements).

Yes, I agree. Should we merge the carousel before next/prev is improved?

Reply:
> Should we merge the carousel before next/prev is improved?

Yes, weapon switching responsiveness should be saved for later, maybe after the next release.

Something else I noticed: Currently the carousel always appears during weapon selection, but with the KEX port, it only appears when next/prev are pressed and not if a weapon is selected directly (e.g. d-pad direction or number keys). Maybe we should match that behavior.

Reply:
Some visual issues:

<details><summary>videos</summary>

Vanilla shotguns:

https://github.com/user-attachments/assets/7fb2eb93-9352-4222-b06a-2e07548749c1

Vanilla fist/saw:

https://github.com/user-attachments/assets/d8b6f89a-0acf-427c-87ce-02434362c4ee

Boom/MBF/MBF21 shotguns:

https://github.com/user-attachments/assets/9055ad22-9f70-44f8-af55-6b3a1a24dc89

Boom/MBF/MBF21 fist/saw:

https://github.com/user-attachments/assets/0582b546-91a7-44bf-9a3e-7b3bc86b99e6

</details>

Recall the earlier discussion [here](https://github.com/fabiangreffrath/woof/issues/1925#issuecomment-2375630591) regarding the need to visualize weapon switching restrictions in a logical manner. Remember this diagram I made of how it's tricky to handle:

![Untitled](https://github.com/user-attachments/assets/1e4a41e0-ad80-41aa-944e-832475989ab1)

Recall how the Half-life games visualize weapon switching. The idea here is similar.

Reply:
Here are some notes I wrote when the KEX port first released. Feel free to address these or ignore anything you think is irrelevant:

<details><summary>notes</summary>

- What happens with wads that change the weapons (e.g. REKKR, Chex)?
  - Show the icons, but they don't match
  - Hide the icons, but now players miss the carousel
  - Wait for community-made icon packs, if that ever happens
- How to incorporate and visualize vanilla switching restrictions?
  - "Next" weapon skips lower priority slots
  - "Prev" weapon includes lower priority slots
- How to implement this without taking up a third of the screen?
  - Single icon in lower right?
  - Draw to a separate texture layer and shrink the scale?
    - But now the UI is a mishmash of different pixel scales

</details>

Reply:
> Some visual issues:

Can we improve it by changing some other `local_player` fields? Or maybe some other variables/structures? The weapon selection code is convoluted.

> Recall how the Half-life games visualize weapon switching. The idea here is similar.

So should we abandon the caruosel and use these icons to create a Half-Life-style widget?

> What happens with wads that change the weapons (e.g. REKKR, Chex)?

This is a problem for modders to fix.

Reply:
> The weapon selection code is convoluted.

I agree.

> So should we abandon the caruosel and use these icons to create a Half-Life-style widget?

No, but I think the carousel needs to be modified visually otherwise it's confusing. The diagram shows "stacked" slots as one possible solution.

> This is a problem for modders to fix.

That's okay, but it's another reason for a menu toggle.

Reply:
> No, but I think the carousel needs to be modified visually otherwise it's confusing. The diagram shows "stacked" slots as one possible solution.

Hmm, then the carousel will really take up a third of the screen. So we need to implement some logic at the `st_carousel.c` level, right? If you can help with this, commits are welcome 😉

Reply:
> If you can help with this, commits are welcome 😉

Maybe hold this for after the next release since it's deceivingly complicated.

Reply:
> When the weapon carousel feature is enabled (it should be a menu toggle

Yes, please. I don't think I would use it.

Reply:
Here's a possible vanilla solution:

<details><summary>vanilla carousel</summary>

![00](https://github.com/user-attachments/assets/bd4366ef-5b0b-42e6-bc8c-a7a0a9c663a0)

Since the icons will overlap if they're very large, the inactive/lower priority weapon can be darkened to help readability:

![04](https://github.com/user-attachments/assets/a7f7bc7e-0e03-44d6-af14-98bf501da5d9)

Here's the same thing with gray boxes to represent the maximum weapon icon dimensions, assumed to be 56x34, and the order in which the icons are drawn. To prevent the shared weapon slot icons from taking up too much screen area, they are overlapped by 50% with the higher priority weapon drawn over the lower priority one.

![01](https://github.com/user-attachments/assets/f0fe2077-9b67-4160-80ea-96b1e3322dda)

If I press "next weapon", the SSG is selected (skips the regular shotgun):

![02](https://github.com/user-attachments/assets/0c12e90d-3242-4194-abc3-a7eb8212a686)

Then if I press "prev weapon", the regular shotgun is selected and drawn over the SSG:

![03](https://github.com/user-attachments/assets/7b331248-7de4-4c2b-8837-29209f9e8153)

The chainsaw and fist would work in a similar way.

</details>

Reply:
> Here's a possible vanilla solution:

How about just darkening disabled icons?
SG disabled:
![woof0001](https://github.com/user-attachments/assets/d220b09f-12ed-4b9b-ae4d-a6e0f84127e0)
SSG selected, SG enabled:
![woof0000](https://github.com/user-attachments/assets/e312f90e-5fb1-4457-acb2-92e8d5a0139e)
 Done in https://github.com/fabiangreffrath/woof/pull/2020/commits/ff6a50d328c8135566765d879a72e40958f10b50

Reply:
> How about just darkening disabled icons?

Sure! Keeps things simple too.

Reply:
> https://github.com/user-attachments/assets/d220b09f-12ed-4b9b-ae4d-a6e0f84127e0

The yellow outlines are missing now.

Reply:
> The yellow outlines are missing now.

Fixed in https://github.com/fabiangreffrath/woof/pull/2020/commits/af42115ee738c63f7baa2101ea9393d05d3c2341